docs/user/build-customization.rst

```diff
@@ -218,3 +223,36 @@ To setup it, you need to define the version of Node.js to use and install the de
          - npm ci
          # Install any other extra dependencies to build the docs
          - npm install -g jsdoc
+
+
+Override the build process completely
+-------------------------------------
+
+.. warning::
+
+   This feature is in a *beta phase* and could suffer incompatible changes or even removed completely in the near feature.
+   It does not yet support some of the Read the Docs' integrations like the :term:`flyout menu`, search and ads.
+   However, integrating all of them is part of the plan.
+   Use it under your own responsibility.
+
+
+When :ref:`extending the build process <build-customization:extend the build process>` is not enough for a particular use case,
+all the default commands can be overwritten to execute only the commands defined by the project on its config file.
+This can be achieve by using :ref:`config-file/v2:build.commands` key as shown in the following example building a `Pelican <https://blog.getpelican.com/>`_ site:
+
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-22.04"
+     tools:
+       python: "3.10"
+     commands:
+       - pip install pelican[markdown]
+       - pelican --settings docs/pelicanconf.py --output output/ docs/
+
+
+As Read the Docs does not have control over the build process,
+you are responsible to run all the commands required to install the requirements and build the documentation properly.
+Once the build process finishes, the ``output/`` folder will be hosted.
```

Comment:
This should be much more obvious in this page. This feels hidden currently. Preferably in bold, and under a heading showing you should _definitely_ read it.

Reply:
It does feel like we're missing some of the requirements here. I feel like we need a list of things. It would help to define the current working directory (repo checkout path), the output path that is synced (`output/`), and might be worth mentioning that `asdf` is used underneath everything too (you probably need to `asdf reshim`)

Reply:
@agjohnson 
> and might be worth mentioning that asdf is used underneath everything too (you probably need to asdf reshim)

It felt weird to me to mention an implementation detail as part of the "usage of our platform". What happens if we stop using `asdf` in the future, or change it for a different tool?

@ericholscher mentioned about avoid adding magic and calling `asdf reshim` after each command and I tend to agree there but...

>> We could execute asdf reshim python after each of the build.commands, for now as well
> This seems like something we should avoid. The whole goal here is to remove magic, so let's not add it back in.

However, as a user, I'd expect that calling `pip install pelican` will immediately make `pelican` executable available in the PATH. Why users will need to know about `asdf` to make usage of `pelican`? Also, note that this is currently a limitation of `asdf` and `python` but it does not happen with `nodejs`. If you do `npm install -g something` that `something` is available in the PATH immediately.

You may think, "Yes, you are adding `-g` to the command to install it globally". However, the global option works differently on `asdf` than on regular `npm` because the command is run as an unprivileged user, and not installed under a global folder. `asdf` is doing some magic underneath to make this work.

I noted in the notes/decisions that `asdf` executes post hooks on particular command names, and `asdf reshim` was automatically executed when `pip` was called. I linked the issue and PR in the PR's description.

That said, I'd like to find a solution that does not involve users knowing anything about `asdf` because that's an implementation detail that they shouldn't care about.

Reply:
Yeah I agree with that, I don't think it serves us well to surface `asdf` internals too much.

> This seems like something we should avoid. The whole goal here is to remove magic, so let's not add it back in.

I feel like we do need to define "magic" that we don't want, as pushing towards "no magic" isn't necessarily what we want. The flavor of magic we don't want is where we assume we know what a user meant to do and do it for them.

In this case, it's probably safe to say the user meant to have the binary available after installing the package, there isn't really another option.

But, I'd agree that a post pip install hook makes the most sense. Executing reshim liberally, after each command, might make it easier to give user some unwanted side effects.

Reply:
`pyenv`, which is used by `asdf` to compile Python versions, is already calling `pyenv-reshim` when it detects `install` or `uninstall` in the command executed: https://github.com/pyenv/pyenv/blob/master/pyenv.d/exec/pip-rehash/pip#L26

(note they have the same for `conda`, `pip` and `easy_install` in the same directory https://github.com/pyenv/pyenv/tree/master/pyenv.d/exec/pip-rehash)

----

`asdf` implemented the pre/post hooks but they had to revert them because it introduced other problems due to the change to not use `exec` anymore: https://github.com/asdf-vm/asdf/issues/475

However, the `asdf-python` plugin has a PR open to wrap the pip command to reshim automatically as `asdf-nodejs` does. This is why it does work for nodejs' `npm` but it does not work for `pip` currently.

We want that PR merged, which will solve our problem. I'll follow up there with them.

Reply:
I went ahead and implemented option 2) for now. The example from `test-builds` without calling `asdf reshim` manually by the user now works out of the box 🎉 

Let me know if you want to change directions here and we can continue talking about this to find the best path at this point.

Reply:
I like option 2 as well :+1: